File: ruby/mergeSort.rb
# マージソート
def margeSort(arr)
    # 内部メソッドだが、スコープはトップレベル
    def m(arr, head, tail) 
        if head < tail
            mid = ((head + tail) / 2).floor
            m(arr, head, mid)
            m(arr, mid + 1, tail)
    
            l = arr.slice(Range.new(head, mid + 1, true))
            r = arr.slice(Range.new(mid + 1, tail + 1, true))
            i = head
            while l.size > 0 && r.size > 0 do
                if l[0] <= r[0]
                    arr[i] = l.shift
                else
                    arr[i] = r.shift
                end
                i += 1
            end
    
            while l.size > 0 do
                arr[i] = l.shift
                i += 1
            end
        end
        return arr
    end
    return m(arr, 0, arr.size-1)
end

arr = [148, 651, 124, 638, 567, 435, 185, 413, 841, 35]
print margeSort(arr)

File: js/mergeSort.js
// マージソート
function margeSort(arr, head, tail) {
  let m = (head, tail) => {
    if (head < tail) {
      let mid = Math.floor((head + tail) / 2);
      m(head, mid);
      m(mid+1, tail);
  
      let l = arr.slice(head, mid+1);
      let r = arr.slice(mid+1, tail+1);
      let i = head;
      while (l.length > 0 && r.length > 0) {
        if (l[0] <= r[0]) {
          arr[i++] = l.shift();
        } else {
          arr[i++] = r.shift();
        }
      }
      while (l.length > 0) {
        arr[i++] = l.shift();
      }
    }
    return arr;
  }

  m(0, arr.length-1);
  return arr;
}
console.log(margeSort([148, 651, 124, 638, 567, 435, 185, 413, 35]));
